src/main.cpp

```diff
@@ -6190,84 +6190,65 @@ bool GetEarliestStakeTime(std::string grcaddress, std::string cpid)
-	{
-		blockhashes += ";" + blockhash;
-		mvCPIDBlockHashes[cpid] = blockhashes;
-	}
+    // Add block hash to CPID hash set.
+    mvCPIDBlockHashes[cpid].insert(blockhash);
 }
 
 StructCPID GetLifetimeCPID(std::string cpid, std::string sCalledFrom)
 {
-	//Eliminates issues with reorgs, disconnects, double counting, etc.. 
-	if (cpid.empty() || cpid=="INVESTOR")
-	{
-		StructCPID stDummy = GetInitializedStructCPID2("INVESTOR",mvResearchAge);
-		return stDummy;
-	}
-	if (fDebug10) printf(" {GLC %s} ",sCalledFrom.c_str());
+    //Eliminates issues with reorgs, disconnects, double counting, etc.. 
+    if (cpid.empty() || cpid=="INVESTOR")
+        return GetInitializedStructCPID2("INVESTOR",mvResearchAge);
+	
+    if (fDebug10) printf(" {GLC %s} ",sCalledFrom.c_str());
 
-
-	std::string hashes = GetCPIDBlockHashes(cpid);
-	std::vector<std::string> vHashes = split(hashes,";");
+    const HashSet& hashes = GetCPIDBlockHashes(cpid);
     ZeroOutResearcherTotals(cpid);
 
-	for (unsigned int i=0; i < vHashes.size(); i++)
-	{
-		std::string myBlockHash = vHashes[i];
-		if (myBlockHash.length() > 5)
-		{
-		    uint256 uHash(myBlockHash);
-			if (!(mapBlockIndex.count(uHash) == 0))
-			{
-				CBlockIndex* pblockindex = mapBlockIndex[uHash];
-				if (pblockindex)
-				{
-					if (pblockindex->IsInMainChain())
-					{
-						if (pblockindex->sCPID == cpid)
-						{
-
-							StructCPID stCPID = GetInitializedStructCPID2(pblockindex->sCPID,mvResearchAge);
-							if (((double)pblockindex->nHeight) > stCPID.LastBlock && pblockindex->nResearchSubsidy > 0)
-							{
-								stCPID.LastBlock = (double)pblockindex->nHeight;
-								stCPID.BlockHash = pblockindex->GetBlockHash().GetHex();
-							}
-							stCPID.InterestSubsidy += pblockindex->nInterestSubsidy;
-							stCPID.ResearchSubsidy += pblockindex->nResearchSubsidy;
-							stCPID.Accuracy++;
-							if (pblockindex->nMagnitude > 0)
-							{
-								stCPID.TotalMagnitude += pblockindex->nMagnitude;
-								stCPID.ResearchAverageMagnitude = stCPID.TotalMagnitude/(stCPID.Accuracy+.01);
-							}
-
-							if (((double)pblockindex->nTime) < stCPID.LowLockTime)  stCPID.LowLockTime  = (double)pblockindex->nTime;
-							if (((double)pblockindex->nTime) > stCPID.HighLockTime) stCPID.HighLockTime = (double)pblockindex->nTime;
-							mvResearchAge[pblockindex->sCPID]=stCPID;
+    for (HashSet::iterator it = hashes.begin(); it != hashes.end(); ++it)
+    {
+        const uint256& uHash = *it;
+
+        // Ensure that we have this block.
+        if (mapBlockIndex.count(uHash) == 0)
+           continue;
+        
+        // Ensure that the block is valid.    
+        CBlockIndex* pblockindex = mapBlockIndex[uHash];
+        if(pblockindex == NULL ||
+           pblockindex->IsInMainChain() == false ||
+           pblockindex->sCPID != cpid)
+            continue;
+
+        // Block located and verified.
+        StructCPID stCPID = GetInitializedStructCPID2(pblockindex->sCPID,mvResearchAge);

```

Comment:
Can we just do this once outside of the loop for the CPID passed to the function?